paasta_tools/check_autoscaler_max_instances.py

```diff
@@ -0,0 +1,159 @@
+            " If so, send an alert if their utilization is above max_instances_alert_threshold."
+        )
+    )
+    parser.add_argument(
+        "-d",
+        "--soa-dir",
+        dest="soa_dir",
+        default=DEFAULT_SOA_DIR,
+        help="Use a different soa config directory",
+    )
+    parser.add_argument(
+        "--dry-run",
+        dest="dry_run",
+        action="store_true",
+        help="Print Sensu alert events instead of sending them",
+    )
+    return parser.parse_args()
+
+
+async def check_max_instances(
+    soa_dir: str,
+    cluster: str,
+    instance_type_class: Type[KubernetesDeploymentConfig],
+    system_paasta_config: SystemPaastaConfig,
+    dry_run: bool = False,
+):
+    kube_client = KubeClient()
+    for service in list_services(soa_dir=soa_dir):
+        service_config = PaastaServiceConfigLoader(service=service, soa_dir=soa_dir)
+        for job_config in service_config.instance_configs(
+            cluster=cluster, instance_type_class=instance_type_class
+        ):
+            instance = job_config.get_instance()
+            if not job_config.get_autoscaling_metric_spec(
+                name=get_kubernetes_app_name(service, instance),
+                cluster=cluster,
+                kube_client=kube_client,
+                namespace=job_config.get_namespace(),
+            ):
+                # Not an instance that uses HPA, don't check.
+                # TODO: should we send status=0 here, in case someone disables autoscaling for their service / changes
+                # to bespoke autoscaler?
+                continue
+
+            if not job_config.get_docker_image():
+                # skip services that haven't been marked for deployment yet.
+                continue
+
+            autoscaling_status = await pik.autoscaling_status(
+                kube_client=kube_client,
+                job_config=job_config,
+                namespace=job_config.get_namespace(),
+            )
+            if autoscaling_status["min_instances"] == -1:
+                log.warning(
+                    f"HPA {job_config.get_sanitised_deployment_name()} not found."
+                )
+                continue
+
+            if (
+                autoscaling_status["min_instances"]
+                == autoscaling_status["max_instances"]
+            ):
+                status = 0
+                output = f"Not checking {service}.{instance} as min_instances == max_instances == {autoscaling_status['max_instances']}. This is probably a canary instance."
+            elif (
+                autoscaling_status["desired_replicas"]
+                >= autoscaling_status["max_instances"]
+            ):
+                threshold = job_config.get_autoscaling_max_instances_alert_threshold()
+                setpoint = job_config.get_autoscaling_params()["setpoint"]
+                metric_threshold_target_ratio = threshold / setpoint
+
+                status = 3
```

Comment:
do we have named constants for these anywhere? i can personally never remember what is CRITICAL and what is WARNING 😆

Reply:
oh, yeah, sensu uses the nagios return codes: https://assets.nagios.com/downloads/nagioscore/docs/nagioscore/3/en/pluginapi.html

Looks like these are defined in pysensu_yelp: https://github.com/Yelp/pysensu-yelp/blob/master/pysensu_yelp/__init__.py#L123